src/gwt/src/org/rstudio/studio/client/workbench/views/terminal/TerminalSessionSocket.java

```diff
@@ -309,15 +326,42 @@ public void dispatchInput(int inputSequence,
       default:
          break;
       }
-      
    }
    
    /**
     * Send output to the terminal emulator.
     * @param output text to send to the terminal
+    * @param detectLocalEcho local-echo detection
     */
-   public void dispatchOutput(String output)
+   public void dispatchOutput(String output, boolean detectLocalEcho)
    {
+      if (detectLocalEcho && !localEcho_.isEmpty())
+      {
+         String lastOutput = "";
+         while (!localEcho_.isEmpty() && lastOutput.length() < output.length())
```

Comment:
I've implemented #1 and #2 which seems to be handling most common command-line editing chores (in Emacs mode). Have a feeling there will be some edge cases to discover/handle but overall it's behaving. #3 may be possible if we discover it's still necessary, but haven't dug into it yet.

Also did some cleanup around checkboxPref (used to create checkboxes bound to a preference for preferences dialog). That call always applied `lessSpaced` to the checkbox, but many callers were also trying to apply other spacings (`spaced`, `tight`, etc.) resulting in multiple conflicting styles that often didn't do what was expected. Added overloads to allow caller to apply style and not emit the default.